---
title: "Analysing the phenotypic spectrum and genetic associations of CLN6 disease with the Human Phenotype Ontology"
author: "Thomas Weissensteiner"
date: "2024-05-22"
output:
  html_document:
    code_folding: show
    toc: true
    toc_float: true
    toc_depth: 4
---
<br> 
<br>

### 1. Motivation and background  
<br>  

#### 1.1. Summarizing phenotypical descriptions at a more general level can generate new insights into the manifestations of a disease

I wrote this script for a paper describing a rare disease caused by genetic variants of the CLN6 gene (1). One aim of the study was analysing the frequency of various symptoms in the patient population, in order to improve guidance for genetic testing.

The [Human Phenotype Ontology (HPO)](https://hpo.jax.org/about) provides an international logical standard to describe and computationally analyse phenotypes characterising human disease (2). Clinical records of 86 patients had been annotated with HPO terms. Of 213 different terms, 31 has been used in five or more patients, and almost all of these referred to nervous system or muscoskeletal symptoms. However, the specificity of the terms varied between patients, and in some cases specific as well as more general terms which included the same symptom(s) had been used.

The script that I wrote served three purposes: 

1. translating the HPO term IDs of the patient records into HPO term names
2. replacing HPO terms of variable specificity with HPO terms at a defined, more general level
3. generating publication-ready plots of HPO term frequencies. 

As major result, **_summarizing HPO terms at the organ system level_** revealed that 23 patients had ocular symptoms, making the **_eye the third most often affected organ in CLN6 disease_**. This was hidden by the heterogeneity of the original HPO annotations, and had not been noted in previous publications.  
<br>  

#### 1.2. How the script summarises Human Phenotype Ontology terms  

The HPO is composed of "parent" and subclass "child" terms, arranged in a directed acyclic graph. "Parent" terms have one or more "children", whereas most "child" terms only have one "parent". However, more complex ancestries exists, as illustrated by this example:

**Fig. 1. Ancestry of the HPO term in "Cherry red spot of the macula**

![](https://github.com/thomas-weissensteiner/portfolio/blob/main/Summarizing_Human_Phenotype_Ontology_terms/Summarizing_Human_Phenotype_Ontology_Terms_figs/HPO_cherryRedSpot_ancestry.jpg)

Several options exist for substituting "Cherry red spot of the macula" with HPO terms of lower specificity at index rank shown in red. 

When the script summarises HPO terms at less specific levels the following decisions are implemented:

* All ancestors of a term with the same index rank will be considered. I.e. in the above example, for term index = 7 "Cherry red spot of the macula" will be summarised as "Abnormal cardiovascular morphology". However, term index ranks 8 or 9 will return both "Abnormal vascular morhoplogy" and "Abnormal posterior eye segment morphology"

* HPO terms with an index rank lower than the chosen one will be preserved

* For an individual patient ID, each summarised term will be counted once, regardless of how many descendants existed in the clinical record

The first two points can be verified by running the code chunk 2 (section 2.1), and looking at the messages that are generated.

Indexing is based on the R package ontologyIndex which is part of the ontologyX suite (3). A recent paper seems to have used the same approach but published no code (4).  
<br>  

### 2. Code for translating HPO IDs to HPO names, and summarising  ontology terms at a common less specific level

Code chunks can be run in R by opening the accompanying  "Summarizing_Human_Phenotype_Ontology_terms.rmd" in RStudio.
The working directory in "Global directory and output options" needs to be customised, and should include the following additional files:
* "variants_scrambled.csv", a tab-delimited csv file in which the original data were randomised
* "Summarizing_Human_Phenotype_Ontology_terms_figs", a folder containing additional figures not generated by the code chunks  
```{r Options, include = FALSE}
## Global directory and output options

knitr::opts_knit$set(root.dir = "C:/Users/ThomasWeissensteiner/OneDrive/Documents/portfolio/Summarizing_Human_Phenotype_Ontology_Terms")
knitr::opts_chunk$set (warning = FALSE)
knitr::opts_chunk$set (tidy = FALSE)
```

```{r Chunk_1, message = FALSE}
# Chunk 1 #
## Load required R libraries and download HPO data

library(tidyverse)
library (ontologyIndex)

## Get latest version of HPO from the The Open Biological and Biomedical Ontologies Foundry
#  Alternative server: http://purl.obolibrary.org/obo/hp.owl

hpo <- get_OBO ("http://purl.obolibrary.org/obo/hp.obo")

## Get example data file

variants_scrambled <- read.delim("https://raw.githubusercontent.com/thomas-weissensteiner/portfolio/main/Summarizing_Human_Phenotype_Ontology_terms/variants_scrambled.csv", quote = "")
```
<br>  


#### 2.1. Contingency table of patient IDs and HPO terms summarised at a defined ontology term level
 
The "findNode" function replaces an HPO term in the original case annotations which new term(s) at the chosen level
Messages are generated whenever  
* an original case annotation was kept because its HPO term was less specific than the chosen term level  
* an HPO term at the chosen level has parent terms in several different branches of the ontology

Setting termIndex_rank to a high value (i.e.., 30) ensures all original HPO annotations are kept.  

```{r Chunk_2, dependson: Chunk_1, attr.output='style="max-height: 100px;"'}
# Chunk 2 requires executing Chunk 1 #

## Set the level at which HPO terms will be summarised 
#  default is 3, in the HPO phenotype branch this corresponds to organ system abnormalities

termIndex_rank = 30


## Function to replace an HPO term in the original case annotations which new term(s) at the chosen level

findNode <- function (x, termIndex_rank, excludeBranches) {
 termsInfo <- get_term_info_content (
 hpo, get_ancestors (hpo, x)) %>%
 .[order (as.numeric(.)) ]

 ancestorNumber <- length (termsInfo)

 if (termIndex_rank <= (ancestorNumber ) ) {
 newTerms <- 
 termsInfo [termsInfo - termsInfo [termIndex_rank]==0] %>% names ()
  if (termsInfo [termsInfo - termsInfo [termIndex_rank]==0] %>% length() > 1) {
    parents <- termsInfo [termsInfo - termsInfo [termIndex_rank]==0] %>% names()
    cat (x, hpo$name [[x]], "summarised by ancestral terms from several HPO branches:", "\n", 
    paste (parents, hpo$name[parents]), "\n")
  }
 return (newTerms)
 } else { paste (x, hpo$name [[x]], "was not summmarised" ) %>% 
 print 
 return (x)
 }
}

## Read and clean data
#  Requires a tab but not comma-delimited cvs file
queryTable <- variants_scrambled %>%	
	  filter (Allele_zygosity == 
	 		"Homozygous" & !HPO_Terms %in% c(NA, "-1", "")) %>%
	  select (ID_person, Gender, disease_.other_name., cDna, Protein,
			Coding_Effect, Clinical_Significance, HPO_Terms)

newList <- list()

for (i in 1 : nrow (queryTable) ) {
queryTerms <- select (queryTable [i, ], HPO_Terms) %>%
separate_rows (., "HPO_Terms", sep=",")
    
newTerms <- apply (queryTerms, 1, findNode, termIndex_rank, infoType)
newRow <- list (unique (newTerms))
    
newList <- c (newList, newRow)
}


## Convert the list of new HPO_Terms to a table in which terms are separated into individual columns
## Ordered by overall frequency of the terms in the table, from left to right
##(a new HPO term which occurs multiple times in a single case is only counted once for that case)

# “termOrder” is a vector of unique new HPO terms, ordered by the number of positive cases

termOrder <- newList %>% lapply (., unlist)%>% lapply (., unique) %>% unlist %>% table() %>% sort (decreasing = T)


# “newTable” stores cases and their new annotations in binary format (1: new HPO term was found at least once, 0: term and its descendants are absent)

newTable <- tibble (queryTable$ID_person)

for (i in 1 : length (termOrder)) {
	newTable <-
    cbind (newTable,			
    ifelse (
    grepl (names(termOrder[i]), newList), 1, 0)
	)	
}

## Reformat contingency table 

# Change column names from HPO identifiers to HPO names

termOrder <- names(termOrder) %>% lapply (., FUN=function (x) { hpo$name[[x]]} ) %>% unlist
newTable <- set_names(newTable, c("ID_person", termOrder))


# Optional: remove level 3 terms that are not an organ system abnormality, e.g. neoplasm or mode of inheritance
# This should be replaced with a filter that can be applied to all levels, i.e. selection of terms in the phenotype branch only

if (termIndex_rank == 3) {
 newTable <- newTable %>%  select (., c(ID_person, grep ("Abnormality", names(.))))
}

newTable <- newTable %>% column_to_rownames(., "ID_person")

## Write the results to a file in the working directory named “HPO. [termIndex_rank] .csv”,
# where [termIndex_rank] is a number indicating the node level of the collapsed HPO_Terms

write.csv (newTable, paste0 ("HPO.level", termIndex_rank, ".csv"))

head (newTable)

```
<br>  


#### 2.2. Barchart of the most frequent HPO terms, summarised at a different levels

##### 2.2.1. HPO terms used in the original case annotations suggest the 10 most frequent diagnoses comprise nervous and muskoskeletal system symptoms  
```{r Chunk_3, dependson: Chunk_2}
# Chunk 3 requires executing Chunk 2 < Chunk 1 #

## Display a barchart of the most frequent new terms
#  summary of HPO term frequencies is generated by newTable %>% ... colSums()  %>% enframe()
#   n sets the number of most frequent terms to be shown, default is slice_head (n=10)   

newTable %>% colSums() %>% enframe() %>%
	slice_head (n=10) %>% 
  ggplot(., aes(x=factor(name, levels = name %>% rev()), y=value)) +
	geom_bar(stat="identity") +
	theme_bw() +
	theme(text=element_text(size=12)) +
	labs (x="HPO Term", y="Number of cases") +
	coord_flip()

```
<br>
<br>  

##### 2.2.2. Frequency of eye abnormalities was masked by the high specificity and heterogeneity of the original HPO terms used for describing ocular symptoms in the patient records

Summarizing HPO terms at the organ system level (termIndex_rank = 3) revealed that the eye is among the third most commonly affected organ


![](https://github.com/thomas-weissensteiner/portfolio/blob/main/Summarizing_Human_Phenotype_Ontology_terms/Summarizing_Human_Phenotype_Ontology_terms_figs/rank3_caseNumbers)


"Abnormality of vision" was the 10th most frequent diagnosis at termIndex_rank = 5


![](https://github.com/thomas-weissensteiner/portfolio/blob/main/Summarizing_Human_Phenotype_Ontology_terms/Summarizing_Human_Phenotype_Ontology_terms_figs/rank5_caseNumbers)  
<br>  
<br>  

### 3.  Exploring associations between gene variants and phenotypes at different HPO term levels 
<br>  

This section contains unpublished work that did not contribute to the CLN6 paper (1). I used two approaches for exploring associations between genetic variants and clinical diagnoses:

a. semantic similarity of the phenotypes (including all HPO terms in the original annotations) among patients carrying the same variant(s)
b. associations between genetic variants and individual HPO terms summarised at different levels

The first approach aims to capture the full information content of the patient records by weighting HPO terms by their specificity (3). It has been used, e.g., to identify new candidate genes based on their association with related diseases (5). However, the clinical phenotype characteristic for a group of specific gene variant(s) is not immediately obvious.
Therefore, I tried a complementary method that tests for association with summarized HPO terms. Here, interpretation of the results should be straightforward. Replacing higher specificity terms with  common ancestors also decreases the number of comparisons, and increases the average size of individual groups. However, this comes with a loss of phenotypic information. Therefore, the term level with the highest statistical power for discriminating specific variant groups might need to be determined experimentally (in a kind of ML way- but I am not statistically astute enough to say how this will affect significance).

The following examples are purely illustrative and do not correspond to results obtained with the actual patient data.  
<br>  


#### 3.1. Identifying variants associated with semantically similar clinical phenotypes

```{r Chunk_4, message = FALSE}
# Chunk 4 #

library(ontologySimilarity)
```
<br>  

##### 3.1.1. Heatmap for exploring variant clustering

Visual inspection can indicate variants or variant groups that are associated with a similar disease phenotype. Original annotations rather than the summarised HPO terms should be used.

Here, phenotypes of patients with variants c.395_396del and c.397_398del cluster together in the upper right corner of the heatmap produced with the original HPO terms of the "variants_scrambled.csv" example file. Another variant associated with semantically similar HPO terms appears to be c.773A>G.

```{r Chunk_5, dependson: Chunk_4, fig.dim = c(13,13)}
# Chunk 5 requires executing Chunk 4 < Chunk 2 < Chunk 1 #

## NOTE: if newList contains summarized HPO terms, run chunck 4 again setting termIndex_rank = 30

## Generate a heatmap for genetic variants, clustered by disease phenotype displayed by their carriers (semantic similarity of sets of HPO terms)


term_sets <- set_names (newList, queryTable$cDna)

sim_mat <- get_sim_grid(ontology=hpo, term_sets=term_sets)

sim_mat %>% 
    .[,order(colnames(.))] %>% 
    heatmap (., scale = "none", 
# Optional: remove "#" to display variants on x-axis in alphabetical order
             # Colv = NA,
             # xlab = "Variants in alphabetical order", 
             # ylab = "Variants clustered by similarity of patient phenotype", 
             cexCol = 0.2, cexRow = 0.2,
             margins = c(5,5)
    )


```
<br>  

##### 3.1.2. Similarity and its p-value for HPO terms associated with selected variant groups

Variant group c.395_396del and c.397_398del and variant c.773A>G are associated with significantly similar phenotypes, whereas c.794_796del and the group of all patients ("All") are not.

```{r Chunk_6, dependson: Chunk_5}
# Chunk6 requires executing Chunk 5 < Chunk4 < Chunk2 < Chunk1 #

## Enter a list of variant groups

group_variants <- list(c("c.395_396del", "c.397_398del"), 
                       "c.773A>G",
                       "c.794_796del",
                       "All")


## Function for generating a table of variant groups and their characteristics

get_similarityTbl <- function(x) {
    groupName <- group_variants[[x]] %>% unlist() %>% paste(., collapse = ", ")
    
    if (groupName == "All") {
        group_index <- 1:nrow(sim_mat)
        n <- nrow(queryTable)
    } else {
        group_index <- which(row.names(sim_mat) %in% unlist(group_variants[[x]]))
        n <- sum(queryTable$cDna %in% unlist(group_variants[[x]]))
    }
    
    if (length(group_index) < 2) {
        stop(paste ("Group", x, "should contain at least two members"))
    }
    
    s <- sim_mat %>% get_sim(group_index)
    p <- sim_mat %>% get_sim_p(group_index)
    
    result <- tibble(
        Group = groupName,
        `n(Group)` = n,
        Similarity = s,
        `p-value` = p
    )
    
    return(result)
}


map_df(seq_along(group_variants), get_similarityTbl)

```
<br>  

#### 3.2. Associations between genetic variants and individual HPO terms summarised at different levels

```{r, CHunk_7, message = FALSE}
# Chunk 7 #

## Load additional R libraries

library(gplots)
library(RColorBrewer)
```

```{r Chunk_8, dependson: Chunk_7, fig.keep='none'}
# Chunk 8 requires executing Chunks 7 < Chunk 2 < Chunk 1) #

## This code is not intended for generating the final plotbut rather dendograms of patient IDs and summarised HPO terms
# The reason is that the distance matrix requires unique column names (provided by the patient IDs)
# Patient IDs are the replaced with variant names (or any other variable from queryTable) afterwards


## Choose methods for distance and clustering methods applied to columns and rows
# rd / rc are for columns, cd / cc for rows because matrix will be transformed later
# Different methods for distance and clustering of person IDs (y - axis) can be explored by changing the "#" ("ward" and "average" should perform best for sparse matrices)


rd <- newTable %>% 
	 dist(., method= "euclidean")
# Change "#" to pick an alternative method for distance of x-axis values
    # dist(., method = # "euclidean")^2
    # as.dist((1-cor(t(.)))/2)
    # dist(., method = "maximum")
    # dist(., method = "manhattan")
    # dist(., method = "canberra")
    # dist(., method = "binary")
    # dist(., method = "minkowski")               

rc<-hclust(rd, method =    "ward.D"           
# Change "#" to pick an alternative method for distance of x-axis clustering
                      # "ward.D2"
                      # "single"
                      # "complete"
                      # "average"
                      # "mcquitty"
                      # "median"
                      # "centroid"
                      )

cd<- newTable %>% 
	t() %>%
	  dist(., method = "euclidean")
# Change "#" to pick an alternative method for distance of y-axis values
  # dist(., method = "euclidean")^2
	# as.dist((1-cor(t(.)))/2)
	# dist(., method = "maximum")
	# dist(., method = "manhattan")
	# dist(., method = "canberra")
	# dist(., method = "binary")
	# dist(., method = "minkowski")                  

cc<-hclust(cd, method =   "ward.D"
# Change "#" to pick an alternative method for distance of y-axis clustering
                     # "ward.D2"
                     # "single"
                     # "complete"
                     # "average"
                     # "mcquitty"
                     # "median"
                     # "centroid"
                     )


##  Generate the dendrogram

hv <- newTable %>%
    data.matrix () %>% t() %>%
    heatmap.2(., scale="none",
            Rowv=as.dendrogram(cc),
Colv=as.dendrogram(rc),
            margins = c(0,0)
    )


## Summarise "loss of function" (LoF) coding effects

queryTable <- queryTable %>% mutate (LoF_status = Coding_Effect %>% str_replace_all (., c("Frameshift"="LoF", "Start_loss"="LoF", "Nonsense"="LoF", "Splicing_mutation"="LoF", "Effect_unknown"="?")))


## Variable to replace patient IDs on the x-axis
# Change "#" to pick an alternative variable

xVar <- "cDna"
      # "LoF_status"

xVar <- pull (queryTable, xVar)
new_labCol <-rep('', nrow(queryTable))
new_labCol[hv$colInd[order(xVar)]] <-	
				xVar[hv$colInd[order(xVar)]]

# Alternative x-axis variable: LoF status
# new_labCol[hv$colInd[order(queryTable$LoF_status)]] <-			queryTable$LoF_status[hv$colInd[order(queryTable$LoF_status)]]


## Variable to be displayed as column side colors
# Change "#" to pick an alternative variable

ColSideVar <-  "LoF_status"
              # "Coding_Effect"
              # "Clinical_Significance"
              # "Gender"

ColSideVar <- pull (queryTable, ColSideVar)

ColSideColors <- brewer.pal (unique(ColSideVar) %>% length(), "Set1") %>%
				factor() %>%
				.[match (ColSideVar, unique(ColSideVar)
						%>% sort()
				)] %>% as.character()

```
<br>  

##### 3.2.1. Variant clustering with the original HPO terms

* Loss of function variants c.395_396del and c.397_398del are associated with a combination of hypotonia and hyperreflexia
* Cerebral atrophy might be more frequent among carriers of the c.773A>G missense variant

```{r, Chunk_9, dependson: Chunk_8, fig.height=12, fig.width=14}
# Chunk 9 > requires executing Chunk 8 < Chunk 7 < Chunk 2 < Chunk 1) #

##  Generate plot with chosen variables for x-axis and optional column side colours


par(cex.main=1)

newTable %>%
    data.matrix () %>% t() %>%
    heatmap.2(., scale="none",
              Rowv=as.dendrogram(cc), 	
              Colv=as.dendrogram(rc),
              col = colorpanel(2, "white", "darkgrey"),
              key = F,
              margins = c(1200/(newList %>% unlist %>% unique %>% length()),36),
              lhei = c(2,12), lwid = c(1,13),
              cexRow = 0.4,
              cexCol = 0.6,
              labCol = new_labCol,
              ColSideColors = ColSideColors,
              tracecol=NA,
              xlab = "Variant",
              ylab = "HPO Term",
              main=paste("Variants and Clinical Presentation")
    )
  legend (y = 1, x = 0.8, xpd=F,
       legend = unique(ColSideVar),
       col = unique(ColSideColors),
       lty= 6,             
       lwd = 5,           
       cex= 0.7
  )
  

```
LoF: "Loss of Function" variants (frameshift, splicing mutation, start_loss, nonsense)
<br>  

```{r, Chunk_10, dependson: Chunk_8, attr.output='style="max-height: 100px;"', warning = TRUE}
# Chunk 10 > requires executing Chunk 8 < Chunk 7 < Chunk 2 < Chunk 1) #

## fix chisq error
## display table in html


## Generate a table showing the number of times each variant is present in different clusters, defined by cutting the dendogram at a height “k”


k = 8
for (i in 1:k) {
	next_cluster <- cutree(rc, k) [hv$colInd] %>%
		{set_names (., queryTable$cDna[hv$colInd])[as.numeric(.)== i]} %>%
		names() %>%
		table() %>%
		enframe(name = "Variant", value = paste ("Cluster", i))
		if (i > 1) {  
			clusters <- full_join (clusters, next_cluster, by = "Variant")
			} else { clusters <- next_cluster	
		}
	}

clusters <- clusters %>% 
column_to_rownames (., var="Variant") %>% 
replace(., is.na(.), 0) %>% 
mutate (Total = rowSums(.)) %>%
.[order(.$Total, decreasing= T ), ] %>%
#	rbind (., Clusters_total=colSums(.) )
	mutate (., 
		Expected = apply (., 2, function(x){
				round(Total*sum(x)/sum(Total), digits = 1)}))

write.csv (clusters, paste0 ("clusters_", k, ".csv"))

clusters

clusters %>% .[ , 1:k] %>% chisq.test()

```
<br>  

##### 3.2.2. Variant clustering with HPO terms summarized at organ system abnormality level

* Carriers of c.395_396del and c.397_398del have a combination of mucoskeletal and nervous system abnormalities, and sometimes one additional affected organ
* all c.773A>G and most c.794_796del and  carriers appear to lack multi-organ involvement, showing only central nervous system symptoms only

![](https://github.com/thomas-weissensteiner/portfolio/blob/main/Summarizing_Human_Phenotype_Ontology_terms/Summarizing_Human_Phenotype_Ontology_terms_figs/rank3_euclDist_WardClust)
<br> 
<br>

### 4. References
  
(1) [Clinical and genetic characterization of a cohort of 97 CLN6 patients tested at a single center](https://doi.org/10.1186/s13023-022-02288-8)
Rus CM, **Weissensteiner T**, Pereira C, ..., Beetz C.
Orphanet J Rare Dis. 2022 May 3;17(1):179

(2) [The Human Phenotype Ontology in 2021](https://doi.org/10.1093/nar/gkaa1043)
Köhler S, Gargano M, Matentzoglu N, ..., Robinson PN.
Nucleic Acids Res. 2021 Jan 8;49(D1):D1207-D1217

(3) [ontologyX: a suite of R packages for working with ontological data](https://doi.org/10.1093%2Fbioinformatics%2Fbtw763)
Greene D, Richardson S, Turro E. 
Bioinformatics. 2017 Apr 1;33(7):1104-1106

(4) [Complex trait associations in rare diseases and impacts on Mendelian variant interpretation](https://doi.org/10.1101%2F2024.01.10.24301111)
Smail C, Ge B, Keever-Keigher MR, Schwendinger-Schreck C, Cheung W, Johnston JJ, Barrett C; Genomic Answers for Kids Consortium; Feldman K, Cohen ASA, Farrow EG, Thiffault I, Grundberg E, Pastinen T.
medRxiv [Preprint]. 2024 Jan 11:2024.01.10.24301111

(5) [Network analysis reveals rare disease signatures across multiple levels of biological organization](https://doi.org/10.1038/s41467-021-26674-1)
Buphamalai P, Kokotovic T, Nagy V, Menche J.
Nat Commun. 2021 Nov 9;12(1):6306
